geyser-plugin-manager/src/geyser_plugin_manager.rs

```diff
@@ -72,4 +52,446 @@ impl GeyserPluginManager {
+
+    #[error("Plugin library path is not specified in the config file")]
+    LibPathNotSet,
+
+    #[error("Invalid plugin path")]
+    InvalidPluginPath,
+
+    #[error("Cannot load plugin shared library")]
+    PluginLoadError(String),
+
+    #[error("The geyser plugin {0} is already loaded shared library")]
+    PluginAlreadyLoaded(String),
+
+    #[error("The GeyserPlugin on_load method failed")]
+    PluginStartError(String),
+}
+
+/// # Safety
+///
+/// This function loads the dynamically linked library specified in the path. The library
+/// must do necessary initializations.
+#[cfg(not(test))]
+pub(crate) fn load_plugin_from_config(
+    geyser_plugin_config_file: &Path,
+) -> Result<(Box<dyn GeyserPlugin>, Library, &str), GeyserPluginManagerError> {
+    use std::fs::File;
+    use std::io::Read;
+
+    let mut file = match File::open(geyser_plugin_config_file) {
+        Ok(file) => file,
+        Err(err) => {
+            return Err(GeyserPluginManagerError::CannotOpenConfigFile(format!(
+                "Failed to open the plugin config file {geyser_plugin_config_file:?}, error: {err:?}"
+            )));
+        }
+    };
+
+    let mut contents = String::new();
+    if let Err(err) = file.read_to_string(&mut contents) {
+        return Err(GeyserPluginManagerError::CannotReadConfigFile(format!(
+            "Failed to read the plugin config file {geyser_plugin_config_file:?}, error: {err:?}"
+        )));
+    }
+
+    let result: serde_json::Value = match json5::from_str(&contents) {
+        Ok(value) => value,
+        Err(err) => {
+            return Err(GeyserPluginManagerError::InvalidConfigFileFormat(format!(
+                "The config file {geyser_plugin_config_file:?} is not in a valid Json5 format, error: {err:?}"
+            )));
+        }
+    };
+
+    let libpath = result["libpath"]
+        .as_str()
+        .ok_or(GeyserPluginManagerError::LibPathNotSet)?;
+    let mut libpath = PathBuf::from(libpath);
+    if libpath.is_relative() {
+        let config_dir = geyser_plugin_config_file.parent().ok_or_else(|| {
+            GeyserPluginManagerError::CannotOpenConfigFile(format!(
+                "Failed to resolve parent of {geyser_plugin_config_file:?}",
+            ))
+        })?;
+        libpath = config_dir.join(libpath);
+    }
+
+    let config_file = geyser_plugin_config_file
+        .as_os_str()
+        .to_str()
+        .ok_or(GeyserPluginManagerError::InvalidPluginPath)?;
+
+    let (plugin, lib) = unsafe {
+        let lib = Library::new(libpath)
+            .map_err(|e| GeyserPluginManagerError::PluginLoadError(e.to_string()))?;
+        let constructor: Symbol<PluginConstructor> = lib
+            .get(b"_create_plugin")
+            .map_err(|e| GeyserPluginManagerError::PluginLoadError(e.to_string()))?;
+        let plugin_raw = constructor();
+        (Box::from_raw(plugin_raw), lib)
+    };
+    Ok((plugin, lib, config_file))
+}
+
+pub fn parse_config_file(
```

Comment:
Get rid of duplicate logic in this function and the above

Reply:
this one is actually unused. will remove